torch/_inductor/kernel/mm.py

```diff
@@ -136,26 +138,64 @@ def tuned_mm(mat1, mat2, *, layout=None):
                 **mm_options(config, m, n, k, layout),
             )
 
-    if use_cutlass_template(layout, m, n, k):
-        CUTLASSGemmTemplate.add_cutlass_gemm_choices(
-            choices, layout, [mat1, mat2], fuseable=True, non_fuseable=True
+    if static_shape and is_nonzero and use_cutlass_template(layout, m, n, k):
+        out_layout = FlexibleLayout(
+            device=layout.device, dtype=layout.dtype, size=layout.size
         )
+        CUTLASSGemmTemplate.add_cutlass_gemm_choices(choices, out_layout, [mat1, mat2])
 
-    from torch._inductor.ir import FixedLayout, FlexibleLayout
+    use_aten = use_aten_gemm_kernels()
 
-    if (
-        len(choices) == 1
-        and use_aten_gemm_kernels()
-        and isinstance(layout, FixedLayout)
-    ):
-        # If we are not autotuning, we can swap to a FlexibleLayout
-        # in order to get fusion optimizations to kick in, e.g. ConcatFusion
-        layout = FlexibleLayout(
-            device=layout.device, dtype=layout.dtype, size=layout.size
-        )
-        choices = [aten_mm.bind((mat1, mat2), layout)]
+    if len(choices) == 0 and not use_aten:
+        log.warning("No choices for GEMM, using ATen backend as fallback")
+        use_aten = True
 
-    return autotune_select_algorithm("mm", choices, [mat1, mat2], layout)
+    if use_aten:
+        choices.extend([aten_mm.bind((mat1, mat2), layout)])
+        if len(choices) == 1 and isinstance(layout, FixedLayout):
+            # If we are not autotuning, we can swap to a FlexibleLayout
+            # in order to get fusion optimizations to kick in, e.g. ConcatFusion
+            out_layout = FlexibleLayout(
+                device=layout.device, dtype=layout.dtype, size=layout.size
+            )
+            choices = [aten_mm.bind((mat1, mat2), out_layout)]
+    try:
+        return autotune_select_algorithm("mm", choices, [mat1, mat2], layout)
+    except NoValidChoicesError:
+        log.warning("All choices for GEMM were invalid, using ATen backend as fallback")
+        choices = [aten_mm.bind((mat1, mat2), layout)]
+        if len(choices) == 1 and isinstance(layout, FixedLayout):
+            # If we are not autotuning, we can swap to a FlexibleLayout
+            # in order to get fusion optimizations to kick in, e.g. ConcatFusion
+            layout = FlexibleLayout(
+                device=layout.device, dtype=layout.dtype, size=layout.size
+            )
+            return aten_mm.bind((mat1, mat2), layout).output_node()
+
+
+def _is_static_problem(inputs_tensors, layout):
+    # checks whether all input tensors and the output layout
+    # have a static shape by attempting to convert the dimensions
+    # to int
```

Comment:
You probably want to use something like the `is_statically_known_list_of_ints` implemented in the `wrapper.py`, which uses the guards information that we already have to simplify the expression.

Reply:
Thanks for the code pointer. I looked it up, and now I'm not sure if that's what I want to use here. I see that the implementation underlying is_statically_known_list_of_ints is this function: https://github.com/pytorch/pytorch/blob/a157a0d00de39fa15f1fd7a5c7c49865bac04f01/torch/_inductor/codegen/wrapper.py#L1527-L1532 which differs from my implementation (just attempt casting to int), by first calling  V.graph._shape_env._maybe_evaluate_static(x) on each dimension and then attempting to cast x to an int. What's surprising to me is that the result of that _maybe_evaluate_static call is just being discarded. I'm not too familiar with the dynamic shape inference system: Does _maybe_evaluate_static have a side effect that would be important here?

Reply:
It doesn't have any side-effect. What it tries to do is use the knowledge from the guards / asserts that the user may have set, etc to evaluate the expression. Sometimes things that look dynamic, may be possibly static after some asserts in the code.

Reply:
Ah, yes, you found a bug there lol

Reply:
I started doubting I understood how this is working ;) I'll prepapre a small fix then...

Reply:
Created https://github.com/pytorch/pytorch/pull/121808